Store modification & deletion dates for series and add range-lookup method to `SeriesService`

LukasKalbertodt opened the pull request:
Like other PRs before, these changes are required by the upcoming Tobira module. The module will enable the new video portal Tobira to synchronize with Opencast somewhat efficiently. It's still in development in https://github.com/elan-ev/opencast-tobira/. We want to push all tobira-unrelated improvements to the community as early as possible though. Hence this PR.

---

To synchronized with Opencast efficiently, Opencast needs to know when a "thing" last changed. That was already true for events (at least in the search service), but Opencast did not know this information for series. This PR changes that by adding a modified date field to series. 

To notify other applications, like Tobira, about a deleted series, we now store a deletion date instead of actually deleting the series.

To use this new information, a new method is added to the `SeriesService` that allows modification date range lookup and returns series sorted by modification date. Unfortunately, to implement the remote impl, we needed to add a new JSON API as well. This is described as internal and is not expected to be used outside of Opencast itself. 

Also see commit message for more details. This PR can probably be reviewed commit-by-commit fairly well.


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

LukasKalbertodt committed:
Add DB index for `modified_date` on `SeriesEntity`

The new `getAllForAdministrativeRead` method performs these date range
queries that would otherwise be very slow.

LukasKalbertodt committed:
Fix and improve error logging in `SeriesServiceDatabaseImpl`

Lots of these were copy & paste bugs. Also, the logger supports
exceptions as last parameter natively. Printing them via `{}` and
`e.getMessage()` is worse.

LukasKalbertodt committed:
Throw `UnauthorizedException` directly from `getAllForAdministrativeRead`

That way it is easier to check the cause of the error instead of having
to unwrap two layers of `SeriesException`s.

mtneug commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:
```diff
@@ -383,6 +403,60 @@ public DublinCoreCatalog getSeries(String seriesId) throws NotFoundException, Se
+    User user = securityService.getUser();
+    if (!user.hasRole(GLOBAL_ADMIN_ROLE) && !user.hasRole(user.getOrganization().getAdminRole())) {
+      throw new UnauthorizedException(user, getClass().getName() + ".getModifiedInRangeForAdministrativeRead");
+    }
```
Does this mean the Tobira API user needs to be (org) admin? I would not give Tobira global admin rights. Org admin maybe, but I'm still unsure if this is a good design. 

Do you expect this method to be used internally or can we simply rely on Spring security for the REST endpoint?

LukasKalbertodt replied:
Currently, we need the Tobira user to be org admin, yes. I was also not happy with this, but ultimately, Tobira needs read access to all events that should potentially appear on the Tobira page. Further, it needs write access to all events that could potentially modified through Tobira. For most institutions that's often simply "Tobira needs read and write access to all videos". After all, users of Tobira perform actions in Opencast *through* Tobira. So it kind of makes sense. This is the same for some other integrations, e.g. the Ilias plugin which also requires a very powerful Opencast user.

Of course, you could always add a special Tobira role to all ACLs to avoid handing out org admin rights. 

Your last quesiton I don't quite understand. I expect the REST endpoint to be used *only* by the series service remote impl.

mtneug replied:
This is not true. ILIAS and Moodle don't require org admin rights and you can thus have videos that are not seen by those systems.

LukasKalbertodt replied:
I just said "a very powerful Opencast user", not org admin. Though yes, that's still not technically correct because the user used by Ilias/Moodle does not *need* to have lots of privileges. But it typically does, i.e. is allowed to read/write the majority (if not all) events.

Again, I also don't like the Tobira module requiring org admin right now. And I would like to change that. I think this can be easily changed in the future. I'm just saying the OC user used by Tobira has very similar requirements to OC users used by Ilias/Moodle.

mtneug replied:
Fair enough and I think we see this similarly :)

LukasKalbertodt commented:
This is still missing an appropriate SQL migration. And I noticed that the code likely has to be adjusted to deal with old series that don't have a valid modification date. So I will work on this and mark as "read for review" as soon as it's done.

LukasKalbertodt committed:
Add SQL migration script for changes to `oc_series`

`deletion_date` and `modified_date` were added. This commit also adds
the corresponding migration script for PostgreSQL since the existing 
script only works/worked for MySQL and MariaDB.

LukasKalbertodt committed:
Add note about DB migration scripts required during upgrade

Arnei commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesEntity.java:
```diff
@@ -49,17 +53,39 @@
-    @NamedQuery(name = "Series.getCount", query = "select COUNT(s) from SeriesEntity s"),
+    @NamedQuery(
+        name = "Series.findAll",
+        query = "select s from SeriesEntity s where s.deletionDate is null"
```
Could you explain to me when do we have a series that is deleted still in the db? As far as I know, deleted series are removed from the db, so storing a deletionDate makes little sense?

LukasKalbertodt replied:
This PR also changes that. When deleting a series, it is not actually deleted from the DB, but we only set the deleted date to a non-null value.

```diff
- em.remove(entity);
+ entity.setModifiedDate(new Date());
+ entity.setDeletionDate(new Date());
+ em.merge(entity);
```

https://github.com/opencast/opencast/pull/3076/commits/afb82d27d61370bb2d5903e4a25881c9aad11401#diff-8ebbecff33097385ffb1bdcc22e6a8acd012f150472562fa17ab98838c7afd43L172-L173

Arnei replied:
Right, not sure how I missed that. Thanks for the link

Arnei replied:
Does that mean there is no way of permanently deleting a series from the db anymore short of directly removing them from the db by hand?

LukasKalbertodt replied:
Yes, that is correct. I mean, for Opencast, it is deleted as all SQL queries retrieving series now explicitly check for `deletion_date is null`. But yes, the data is still in the database. 

Maybe instead of just setting the modification and deletion dates on removal, all other columns should be set to empty or dummy values? But that could always be changed later I guess?

Arnei commented:
Looks good to me, but I think I'd like a second opinion here. Particularly on the "series aren't actually deleted anymore"-part.  So far, I've not actually managed to come up with a good argument on why this would be a problem, but I still feel like it could be a problem. Maybe there are some database limitations or weird data privacy laws I'm unaware of.

lkiesow commented on docs/upgrade/10_to_11/postgresql9.sql:
```diff
@@ -0,0 +1,23 @@
+UPDATE oc_series
+    SET modified_date = TIMESTAMP '1970-01-01 00:00:01'
+    WHERE modified_date IS NULL;
+ALTER TABLE oc_series ALTER COLUMN modified_date SET NOT NULL;
```
Does it make sense to run at the end:

```sql
VACUUM (VERBOSE, ANALYZE) oc_series;
```

LukasKalbertodt replied:
Not sure to be honest. At least in postgres (as far as I know), the first 8 nullable columns don't have any overhead as there is always an 8bit mask per row for exactly that. But :shrug:

lkiesow commented on modules/dublincore/src/main/java/org/opencastproject/metadata/dublincore/DublinCoreGsonAdapter.java:
```diff
@@ -0,0 +1,52 @@
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import org.json.simple.parser.ParseException;
```
If you use Gson, you probably also want to use Gson's exception type:
[`com.google.gson.JsonParseException](https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/JsonParseException.html)

LukasKalbertodt replied:
Well, I am not throwing that exception. I am catching it. Because I'm using an existing method which uses JSON simple. It's a bit stupid but was the easiest solution. Everything else would have required writing lots more code and likely duplicating logic. Kinda annoying to have multiple JSON libraries.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java:
```diff
@@ -345,6 +350,91 @@ public Response getSeriesAccessControlListJson(@PathParam("seriesID") String ser
+          Optional.ofNullable(to).map(millis -> new Date(millis)),
+          limit);
+
+      Gson gson = new Gson();
```
Gson is thread-safe, and you can make this a class field. Constructing this every time has quite some overhead and caused problems for Opencast at several places.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java:
```diff
@@ -345,6 +350,91 @@ public Response getSeriesAccessControlListJson(@PathParam("seriesID") String ser
+   * message as body. Also logs the message.
+   */
+  private static Response badRequest(String msg) {
+    logger.warn("Bad request to /series/allInRangeAdministrative: {}", msg);
```
This lets users spam the logs with warnings.
For a client error, the error is returned.
Make this a debug log:
```suggestion
    logger.debug("Bad request to /series/allInRangeAdministrative: {}", msg);
```

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:
```diff
@@ -169,12 +177,15 @@ public void deleteSeries(String seriesId) throws SeriesServiceDatabaseException,
       }
-      em.remove(entity);
+
+      entity.setModifiedDate(new Date());
+      entity.setDeletionDate(new Date());
```
Maybe make sure they are actually identical?
```suggestion
      var date = new Date();
      entity.setModifiedDate(date);
      entity.setDeletionDate(date);
```

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:
```diff
@@ -383,6 +403,60 @@ public DublinCoreCatalog getSeries(String seriesId) throws NotFoundException, Se
+            .setParameter("to", to.get())
+            .setMaxResults(limit);
+      } else {
+        q = em.createNamedQuery("Series.getAllModifiedSince", SeriesEntity.class)
```
Above, you check if the user is an admin or an organizational admin. This query returns all events from all organizations. This means, you allow admins from one organization to access events from another organization. I don't think that's intended, is it?

lkiesow replied:
This is now resolved in the query `getAllModifiedSince`

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:
```diff
@@ -383,6 +403,60 @@ public DublinCoreCatalog getSeries(String seriesId) throws NotFoundException, Se
+      }
+
+      final List<Series> out = new ArrayList<>();
+      for (SeriesEntity entity : q.getResultList()) {
```
`SeriesEntity` holds basically the same data the new `Series` does. Would it make sense to have `SeriesEntity` implement/extend `Series`?

LukasKalbertodt replied:
I don't really think so. `SeriesEntity` for me is just the connection to the DB and I would not want to mix it with `Series`. For example, `Series` contains `DublinCoreCatalog dublinCore`. If `SeriesEntity` would derive from `Series` that would have to be `null` all the time? And from experience, in these situations, sooner or later there will be more and more differences between the "database adapter" and the thing you want to use in your actual code.

lkiesow replied:
I would have probably solved that by making `Series` an interface and just requiring the methods, but fair enough, it's not really important.

lkiesow commented on modules/series-service-remote/src/main/java/org/opencastproject/series/remote/SeriesServiceRemoteImpl.java:
```diff
@@ -322,6 +332,47 @@ public DublinCoreCatalog getSeries(String seriesID) throws SeriesException, NotF
+        throw new UnauthorizedException("got UNAUTHORIZED when fetching /allInRangeAdministrative.json");
+      } else {
+        // Retrieve and deserialize data
+        Gson gson = new Gson();
+        Type type = new TypeToken<ArrayList<Series>>() { }.getType();
```
Make these class fields.

lkiesow commented on modules/series-service-impl/src/main/java/org/opencastproject/series/endpoint/SeriesRestService.java:
```diff
@@ -345,6 +350,91 @@ public Response getSeriesAccessControlListJson(@PathParam("seriesID") String ser
+          @RestParameter(
+              name = "to",
+              isRequired = false,
+              // TODO: this shows the default value as 0 despite us not setting this value!
```
```suggestion
              defaultValue = "",
```

LukasKalbertodt replied:
Nop, that still shows `0` in the test form box :/

lkiesow replied:
Interesting. Well… let's ignore that for now.

LukasKalbertodt committed:
Add proper `DublinCoreCatalog` (de)serialization when using Gson

The default Gson serialization with this class is rather bad and the
deserialization actually fails. Instead, we now use the already existing
methods for JSON (de)serialization.

LukasKalbertodt committed:
Store deletion and modified date for series in the DB

This is useful for several things, but in particular: we need this for
the upcoming Tobira harvesting API. Without information about when
something about a series changes, a syncing application would have to
request all series every time, which is quite a waste.

To also be able to notify other applications about deleted series, we
cannot delete them for real, but just flag them as deleted. Instead of
a flag, a timestamp is stored as that might be useful in the future.

LukasKalbertodt committed:
Add `SeriesService.getAllForAdministrativeRead`

This method will be used in the Tobira harvesting API to retrieve all
series in a specific time range. Since the dublin core catalogue is not
sufficient for several use cases, a new class `Series` is added as
well.

LukasKalbertodt committed:
Only return series belonging to the current organization in getAllForAdministrativeRead

LukasKalbertodt committed:
Properly implement `getAllForAdministrativeRead` for series remote impl

This was left out from the initial commit adding this method. It uses
the JSON endpoint added last commit.

LukasKalbertodt committed:
Add series JSON endpoint `allInRangeAdministrative.json`

This endpoints only purpose is to implement `getAllForAdministrativeRead`
for the remote impl. It is an internal API and should not be used for
any other purpose.

LukasKalbertodt commented:
@lkiesow I think I fixed all of your remarks (except the four which I comment-replied to). Most changes have been amended to the respective commits. Only the security fix is its own commit.

KatrinIhler commented on modules/series-service-impl/src/main/java/org/opencastproject/series/impl/persistence/SeriesServiceDatabaseImpl.java:
```diff
@@ -303,13 +316,20 @@ public DublinCoreCatalog storeSeries(DublinCoreCatalog dc)
+        // If the series existed but is marked deleted, we completely delete it
+        // here to make sure no remains of the old series linger.
+        if (entity != null) {
+          this.deleteSeries(seriesId);
```
Just ran into this while testing my migration scripts for Bern - this doesn't actually seems to work, since deleteSeries() would only mark the entity as deleted again, not remove it (found no method to do this), but mostly because this throws a NotFound for entities which are already marked as deleted. To test this, just create a series, delete it, and then try to recreate a series with the same id. Since you know how this feature is supposed to work, can you take a look at this?
